### PR TITLE
Multivariate floating point precision fix

### PR DESCRIPTION
**Pull request by ericmjl**
This is a bug fix for `distributions.multivariate`'s Multinomial errors related to floating point precision. This will very likely be helpful for computation runs that involve the GPU, and involve many classes with small floating point numbers.

More information can be found here: https://github.com/pymc-devs/pymc3/issues/2469

### CODE REVIEW

**Comment by ericmjl** [comment]
Just saw a test fail, I think I know what's going on. Let me fix that.

**Comment by ericmjl** [comment]
Okay, locally tests seem like they're passing.

```bash
$  py.test pymc3/tests/test_distributions_random.py
============================ test session starts ============================
platform linux -- Python 3.6.0, pytest-3.0.5, py-1.4.32, pluggy-0.4.0
rootdir: /home/ericmjl/github/software/pymc3, inifile: setup.cfg
collected 214 items

pymc3/tests/test_distributions_random.py ....................................
.............................................................................
.............................................................................
.....................xss

============ 211 passed, 2 skipped, 1 xfailed in 253.00 seconds =============

$ py.test pymc3/tests/test_distributions.py
============================ test session starts ============================
platform linux -- Python 3.6.0, pytest-3.0.5, py-1.4.32, pluggy-0.4.0
rootdir: /home/ericmjl/github/software/pymc3, inifile: setup.cfg
collected 95 items

pymc3/tests/test_distributions.py ....................................x...x..
...xxx...x.....................................x....

=================== 88 passed, 7 xfailed in 55.32 seconds ===================
```

I'm feeling a bit more confident now that the tests on Travis should pass.

**Comment by ericmjl** [comment]
@ColCarroll @junpenglao all tests pass!!!!!!

File: pymc3/distributions/multivariate.py [hunk 1]
```diff
@@ -503,17 +503,20 @@ def __init__(self, n, p, *args, **kwargs):
         self.mode = tt.cast(tround(self.mean), 'int32')
 
     def _random(self, n, p, size=None):
+        # Set float type to float64 for numpy.
```
Comment (ColCarroll) [inline pymc3/distributions/multivariate.py]: can you add a comment with a link to the `numpy` issue -- that'll make it clearer why this cast exists, and if/when we can remove it

File: pymc3/distributions/multivariate.py [hunk 2]
```diff
@@ -503,17 +503,20 @@ def __init__(self, n, p, *args, **kwargs):
...
+            p = p / (p.sum())
             randnum = np.random.multinomial(n, p.squeeze(), size=size)
         elif p.ndim == 2:
+            p = p / (p.sum(axis=1, keepdims=True))
             randnum = np.asarray([np.random.multinomial(n, pp, size=size) for pp in p])
         else:
             raise ValueError('Outcome probabilities must be 1- or 2-dimensional '
                              '(supplied `p` has {} dimensions)'.format(p.ndim))
-
         return randnum
```
Comment (ColCarroll) [inline pymc3/distributions/multivariate.py]: should the return value get cast back to the original data type?  honest question -- maybe @kyleabeauchamp has an opinion?

File: pymc3/distributions/multivariate.py [hunk 3]
```diff
@@ -503,17 +503,20 @@ def __init__(self, n, p, *args, **kwargs):
...
 
     def _random(self, n, p, size=None):
+        # Set float type to float64 for numpy.
+        p = p.astype('float64')
+        # Now, re-normalize all of the values in float64 precision.
         if size == p.shape:
             size = None
-
         if p.ndim == 1:
+            p = p / (p.sum())
```
Comment (ColCarroll) [inline pymc3/distributions/multivariate.py]: nit -- extra parens around `p.sum(...)`, both here and below

**Review by ColCarroll — COMMENTED** [review]
looks good to merge now -- had a few small comments if you have a second.  excited to try this out and see if it fixes my problems!

File: pymc3/distributions/multivariate.py (hunk 2 above)
Comment (kyleabeauchamp) [reply pymc3/distributions/multivariate.py]: No idea.  But X is discrete while p is continuous, so presumably if it worked previously so it should continue working with the current patch.

File: pymc3/distributions/multivariate.py (hunk 1 above)
Comment (ericmjl) [reply pymc3/distributions/multivariate.py]: No problem, @ColCarroll!

File: pymc3/distributions/multivariate.py (hunk 3 above)
Comment (ericmjl) [reply pymc3/distributions/multivariate.py]: Fixed, going to push up.

File: pymc3/distributions/multivariate.py (hunk 2 above)
Comment (ericmjl) [reply pymc3/distributions/multivariate.py]: I think it's a good idea. I will implement that fix.

**Comment by ericmjl** [comment]
@ColCarroll I ran the two relevant tests locally and found that they pass. I'm confident it'll pass on Travis, but as usual, let's wait it out.

**Comment by ericmjl** [comment]
@ColCarroll @junpenglao @kyleabeauchamp I think one of the tests timed out, is that correct? Otherwise, everything else has passed.

**Comment by ColCarroll** [comment]
Yeah, one is a flaky test, one looks like coveralls threw a 5xx error.  I restarted both.

**Comment by ColCarroll** [comment]
Thanks for fixing this!

**Comment by ericmjl** [comment]
Yay!!!! Thanks @ColCarroll!

**Pull request merged**